drivers: modem: fix uninitialized variable warnings in modem_stats

ct-fk opened the pull request:
The variables 'first' and 'next' in the functions 'stats_buffer_list_first()' and
'stats_buffer_list_next()' were potentially used uninitialized. Depending on the
compiler and target architecture, this can lead to different behavior, including
warnings or errors when using strict warning flags. By initializing these pointers
to 'NULL', we ensure consistent and expected behavior across all toolchains and
architectures.

Signed-off-by: Fabian Kainka <kainka@cognid.de>

tomi-font reviewed (changes requested):
No?

tomi-font commented:
I mean, did you actually get some error because of those? The one at line 113 would be very surprising as `buffer` is assigned right after the variable declarations.

ct-fk commented:
> I mean, did you actually get some error because of those? The one at line 113 would be very surprising as `buffer` is assigned right after the variable declarations.

You are right about line 113, but I encountered the following errors during compilation:

/home/fk/Dokumente/workspace/modular_nrf/zephyr/subsys/modem/modem_stats.c:41:16: error: 'first' may be used uninitialized [-Werror=maybe-uninitialized]
/home/fk/Dokumente/workspace/modular_nrf/zephyr/subsys/modem/modem_stats.c:52:16: error: 'next' may be used uninitialized [-Werror=maybe-uninitialized]

These errors were produced using arm-zephyr-eabi-gcc targeting a Cortex-M33 MCU, with the -Werror flag treating warnings as errors. Given that Zephyr supports many different toolchains and optimization settings, I believe it's safer to explicitly initialize these variables to NULL to prevent similar warnings for other developers using different compiler settings.

ct-fk committed:
drivers: modem: initialize variables to avoid warn

The variables 'first' and 'next' in function 'stats_buffer_list_first()'
and 'stats_buffer_list_next()' were potentially used uninitialized.

Depending on the compiler and target architecture, this can lead to
different behavior, including warnings or errors when using strict
warning flags.

By initializing these pointers to 'NULL', we ensure consistent and
expected behavior across all toolchains and configurations.

Signed-off-by: Fabian Kainka <kainka@cognid.de>

tomi-font reviewed (approved):
Yeah you're right. Thanks for the fix!

bjarki-andreasen commented:
The only reason I can see for any compiler to complain here is that it can't understand the `K_SPINLOCK()` macro, thinking the code block is only optionally executed. Sucks to add unnecessary code just to please compilers (initializing the variable is not free)
